### PR TITLE
[CHIA-1427]: Limit full file creation when processing subscription generations

### PR DESCRIPTION
**Pull request by emlowe**
When syncing a subscription, DL downloads the delta files for each generation as it walks all the singleton generations. Currently each time it will also write out a full file - however, that full file isn't used and will be eventually deleted as the default is to only keep on full file.

This PR is an attempt to limit creating the full files except when we are within the range where we should generate the full files. 

It will also create a full file if a delta file fails to download entirely - this makes sure that if in the middle of processing the deltas the code happens to be unable to move forward, a full file will get generated at that point.

### CODE REVIEW

File: chia/data_layer/download_data.py [hunk 1]
```diff
@@ -92,7 +92,8 @@ async def insert_into_data_store_from_file(
     store_id: bytes32,
     root_hash: Optional[bytes32],
     filename: Path,
-) -> None:
+) -> int:
+    num_inserted = 0
```
Comment (altendky) [inline chia/data_layer/download_data.py:96]: This looks like debugging noise from the other stuff you were figuring out?  Seems reasonable to have, just making sure I'm not missing a relationship to the core change here.

File: chia/data_layer/download_data.py [hunk 2]
```diff
@@ -283,21 +288,25 @@ async def insert_from_delta_file(
...
                 f"Successfully inserted hash {root_hash} from delta file. "
-                f"Generation: {existing_generation}. Store id: {store_id}."
+                f"Generation: {existing_generation}. Store id: {store_id}. Nodes inserted: {num_inserted}."
             )
 
-            root = await data_store.get_tree_root(store_id=store_id)
-            with open(filename_full_tree, "wb") as writer:
-                await data_store.write_tree_to_file(root, root_hash, store_id, False, writer)
-            log.info(f"Successfully written full tree filename {filename_full_tree}.")
+            if target_generation - existing_generation <= maximum_full_file_count - 1:
```
Comment (altendky) [inline chia/data_layer/download_data.py:302]: maybe?  or just mark this conversation as resolved.

```suggestion
            if target_generation - existing_generation < maximum_full_file_count:
```

File: chia/data_layer/data_layer.py [hunk 3]
```diff
@@ -621,6 +624,30 @@ async def fetch_and_validate(self, store_id: bytes32) -> None:
             except Exception as e:
                 self.log.warning(f"Exception while downloading files for {store_id}: {e} {traceback.format_exc()}.")
 
+        # if there aren't any servers then don't try to write the full tree
+        if not success and len(servers_info) > 0:
```
Comment (altendky) [inline chia/data_layer/data_layer.py:628]: Should we just entirely short circuit with a `return` above if there are no servers and skip the loop etc?

File: chia/data_layer/data_layer.py [hunk 4]
```diff
@@ -568,6 +568,7 @@ async def fetch_and_validate(self, store_id: bytes32) -> None:
         servers_info = await self.data_store.get_available_servers_for_store(store_id, timestamp)
         # TODO: maybe append a random object to the whole DataLayer class?
         random.shuffle(servers_info)
+        success = False
```
Comment (altendky) [inline chia/data_layer/data_layer.py:571]: I think if we change the `break` on success to a `return` then we don't need this to be a 'bigger scoped' variable.

**Review by altendky — CHANGES_REQUESTED** [review]
Seems like a good change to me.  I put in a few code form comments, but I'm not strongly attached to them.

File: chia/data_layer/download_data.py (hunk 1 above)
Comment (emlowe) [reply chia/data_layer/download_data.py:96]: Yea, it's somewhat unrelated but I added some logging that reports how many nodes were inserted when we inserted this delta file. Probably of some limited value but I don't think hurts anything so inclined to leave it in

**Review by altendky — APPROVED** [review]
comments are being considered separately for potential refactoring later

**Review by markelrod — APPROVED** [review]

**Pull request merged**